### PR TITLE
e2e latency fix

### PR DESCRIPTION
**Pull request by dadlex**

### CODE REVIEW

File: src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemTimestampSerializer.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	implements Serializer<I>, Serializable {
+
+	private final boolean useDirectMem;
+
+	/**
+	 * @param useDirectMem Specifies whether should it use direct memory for the resulting buffer containing the
+	 *                     serialized data or not. Using the direct memory may lead to the better performance in case of
+	 *                     large data chunks but less safe.
+	 */
+	public DataItemTimestampSerializer(final boolean useDirectMem) {
```
Comment (dadlex) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemTimestampSerializer.java]: We don't know inside serializer if we should add timestamp or no, so a new serializer is added. I'd like to hear suggestions if there are any.

Comment (akurilov) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemTimestampSerializer.java]: just add one more flag arg

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 2]
```diff
@@ -133,6 +136,7 @@
 	private final boolean controlScopeFlag;
 	private final boolean controlStreamFlag;
 	private final Credentials cred;
+	private int timestampLength = 8;
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: static final USE_CAPITAL_LETTERS_FOR_CONSTANTS

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 3]
```diff
@@ -893,6 +902,20 @@ void readEvent(ReaderGroupManager readerGroupManager,String evtReaderGroupName,
 		evtOp.finishResponse();
 		val evtRead = evtRead_;
 		val evtData = evtRead.getEvent();
+		if (tailReadFlag) {
+			val timestampBuffer = ByteBuffer.allocate(timestampLength);
+			timestampBuffer.put(evtData.array(),0,8);
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: magic number

**Review by akurilov — DISMISSED** [review]

**Comment by akurilov** [comment]
Where's the corresponding usage documentation?

File: src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemTimestampSerializer.java (hunk 1 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemTimestampSerializer.java]: that's how it works in this PR

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 3 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: \+

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 2 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: \+

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 4]
```diff
@@ -177,7 +184,6 @@ public final StreamConfiguration apply(final String scopeName) {
 			final StreamConfiguration streamConfig = StreamConfiguration
 				.builder()
 				.scalingPolicy(scalingPolicy)
-				.scope(scopeName)
```
Comment (AndreyKoltsov1997) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:181]: Am I right it was an initial scope name and, removing this, we're leaving the name generation to Pravega?

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 5]
```diff
@@ -316,10 +322,18 @@ public PravegaStorageDriver(
 		super(stepId, dataInput, storageConfig, verifyFlag, BYTES.equals(streamDataType) ? 1 : batchSize);
 		this.concurrencyThrottle = new Semaphore(concurrencyLimit > 0 ? concurrencyLimit : Integer.MAX_VALUE, true);
 		val driverConfig = storageConfig.configVal("driver");
+		val createConfig = driverConfig.configVal("create");
+		this.recordWriteTimeFlag = createConfig.boolVal("timestamp");
+		if (recordWriteTimeFlag) {
+			Loggers.ERR.warn("{}: Make sure that event size is not smaller than 8 bytes ", stepId);
```
Comment (AndreyKoltsov1997) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: Could you, please, explain in a high level, what'd happen it the event size would be smaller than 8 bytes?
If things would be broken, maybe it'd be more sufficient to throw an exception or change log level to error?

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 6]
```diff
@@ -897,23 +922,27 @@ void readEvent(ReaderGroupManager readerGroupManager,String evtReaderGroupName,
 		evtOp.finishResponse();
 		val evtRead = evtRead_;
 		val evtData = evtRead.getEvent();
+		if (tailReadFlag) {
+			val timestampBuffer = ByteBuffer.allocate(TIMESTAMP_LENGTH);
+			timestampBuffer.put(evtData.array(),0,TIMESTAMP_LENGTH);
```
Comment (AndreyKoltsov1997) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: A really minor notice: this comma separation style is differ from other code - it misses white space. I'd consider to add it.

File: src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemSerializer.java [hunk 7]
```diff
@@ -15,14 +15,15 @@
...
 	private final boolean useDirectMem;
-
+	private final boolean shouldRecordTime;
 	/**
 	 * @param useDirectMem Specifies whether should it use direct memory for the resulting buffer containing the
 	 *                     serialized data or not. Using the direct memory may lead to the better performance in case of
 	 *                     large data chunks but less safe.
 	 */
-	public DataItemSerializer(final boolean useDirectMem) {
+	public DataItemSerializer(final boolean useDirectMem, final boolean shouldRecordTime) {
```
Comment (AndreyKoltsov1997) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemSerializer.java:25]: I'd consider adding doc string for shouldRecordTime parameter.

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java [hunk 8]
```diff
@@ -845,9 +859,19 @@ void readEvents(final List<O> evtOps) {
...
 				val readerGroupConfig = evtReaderGroupConfigCache.computeIfAbsent(
-					scopeName + SLASH + streamName, key -> readerGroupConfigBuilder.stream(key).build());
+					scopeName + SLASH + streamName, key -> {
+							StreamManager streamManager = StreamManager.create(clientConfig);
+							StreamInfo streamInfo = streamManager.getStreamInfo(scopeName, streamName);
+							StreamCut streamCut = tailReadFlag ? streamInfo.getTailStreamCut() : streamInfo.getHeadStreamCut();
+							return readerGroupConfigBuilder
+								.stream(Stream.of(scopeName, streamName),
+									streamCut,
+									StreamCut.UNBOUNDED)
```
Comment (AndreyKoltsov1997) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:871]: Am I right you're building a config for reader group which readers would read events from Stream.of(scopeName, streamName) up to streamCut?

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 4 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:181]: it was a deprecated method in 0.6 and it's completely removed in 0.7. Stream creation requires scope name, stream name and stream config. So, this parameter is excessive now.

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 5 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: Mongoose would crash on write operation. Throwing an exception would be nice but the size of event is controlled by `item-data-size` which is handled in mongoose-base and we don't know its value in driver and we shouldn't know. I'll raise level of this log.

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 6 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java]: agreed

File: src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java (hunk 8 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java:871]: .stream() takes 3 parameters to configure a reader group: 1) scope/stream to read from, 2) left-border streamCut to read from, 3) right-border streamCut to read to. Here we set left border either to start of the stream or the end and we don't set any restrictions on the right border.

File: src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemSerializer.java (hunk 7 above)
Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/io/DataItemSerializer.java:25]: will do

File: src/main/java/com/emc/mongoose/storage/driver/pravega/EndToEndLogMessage.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package com.emc.mongoose.storage.driver.pravega;
+
+import com.emc.mongoose.base.logging.LogMessageBase;
+import static java.lang.System.lineSeparator;
+
+public class EndToEndLogMessage
+	extends LogMessageBase {
```
Comment (AndreyKoltsov1997) [inline src/main/java/com/emc/mongoose/storage/driver/pravega/EndToEndLogMessage.java:7]: Strange formatting - there's a lot more strings that're longer below.

Comment (dadlex) [reply src/main/java/com/emc/mongoose/storage/driver/pravega/EndToEndLogMessage.java:7]: Generally we start a new line when reach "extends" keyword

File: src/test/java/com/emc/mongoose/storage/driver/pravega/integration/PravegaEventReadTest.java [hunk 10]
```diff
@@ -35,7 +36,8 @@ public void testEventRead()
 						.build();
 		streamManager.createStream(scopeName, streamName, streamConfig);
 		try (
-						val clientFactory = ClientFactory.withScope(scopeName, controllerURI);
+						val clientFactory = EventStreamClientFactory.withScope(scopeName,
+								ClientConfig.builder().controllerURI(controllerURI).build());
```
Comment (AndreyKoltsov1997) [inline src/test/java/com/emc/mongoose/storage/driver/pravega/integration/PravegaEventReadTest.java]: Shouldn't parameters listed from new line be at least 1 tab away from instance / class name? Like it does here: https://github.com/emc-mongoose/mongoose-storage-driver-pravega/blob/master/src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java#L532

File: src/test/java/com/emc/mongoose/storage/driver/pravega/integration/PravegaEventReadTest.java [hunk 11]
```diff
@@ -56,7 +58,8 @@ public void testEventRead()
 			readerGroupManager.createReaderGroup(readerGroup, readerGroupConfig);
 		}
 		try (
-						val clientFactory = ClientFactory.withScope(scopeName, controllerURI);
+				val clientFactory = EventStreamClientFactory.withScope(scopeName,
+						ClientConfig.builder().controllerURI(controllerURI).build());
```
Comment (AndreyKoltsov1997) [inline src/test/java/com/emc/mongoose/storage/driver/pravega/integration/PravegaEventReadTest.java:61]: Same for the tab intent. Shouldn't it be like that: https://github.com/emc-mongoose/mongoose-storage-driver-pravega/blob/master/src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java#L532 ?

**Review by AndreyKoltsov1997 — APPROVED** [review]
The logic in the code LGTM.
I'd highly recommend consider defining a unified code style for the driver.

**Pull request merged**